apps/prairielearn/src/sync/fromDisk/authors.ts

```diff
@@ -0,0 +1,141 @@
+    resolved.forEach((course, sharingName) => {
+      if (course) {
+        sharingNameLookupTable.set(sharingName, course.id);
+      } else {
+        // This should never happen since all sharing names should have been validated before
+        throw new Error(`Course with sharing name ${sharingName} not found!`);
+      }
+    });
+  }
+  return sharingNameLookupTable;
+}
+
+export async function sync(
+  courseId: string,
+  courseData: CourseData,
+  questionIds: Record<string, string>,
+) {
+  if (infofile.hasErrors(courseData.course)) {
+    return;
+  }
+
+  interface JSONAuthor {
+    name?: string;
+    email?: string;
+    orcid?: string;
+    originCourse?: string;
+  }
+
+  interface NormalizedAuthor {
+    name: string | null;
+    email: string | null;
+    orcid: string | null;
+    origin_course: string | null;
+    id: string | null;
+  }
+
+  // Collect all unique authors from all questions
+  // Also setting up reverse lookup of resolved authors from database back to JSON
+  const uniqueAuthors = new Map<string, JSONAuthor>();
+  // Sharing name -> course ID mappings (resolved in bulk)
+  const sharingNameLookupTable = await resolveSharingNames(courseData);
+
+  for (const qid of Object.keys(courseData.questions)) {
+    const question = courseData.questions[qid];
+    if (infofile.hasErrors(question)) continue;
+    const authors = question.data?.authors ?? [];
+    for (const author of authors) {
+      const normalizedOrcid = author.orcid?.replaceAll('-', '');
+      const normalizedAuthor: NormalizedAuthor = {
+        name: author.name ?? null,
+        email: author.email ?? null,
+        orcid: normalizedOrcid ?? null,
+        origin_course: author.originCourse
+          ? (sharingNameLookupTable.get(author.originCourse) ?? null)
+          : null,
+        id: null,
+      };
+
+      uniqueAuthors.set(JSON.stringify(normalizedAuthor), author);
+    }
+  }
+
+  const authorsForDB = {
+    authors: JSON.stringify(Array.from(uniqueAuthors.keys()).map((a) => JSON.parse(a))),
+  };
+
+  // Insert all unique authors
+  // In the SQL, we ensure that only non-existing authors are actually, so keeping the entire list is fine here
+  await sqldb.execute(sql.insert_authors, authorsForDB);
+
+  // Re-load authors from DB (including IDs) and build new map directly from JSONAuthor to ID
+  const authors = await sqldb.queryRows(sql.select_authors, authorsForDB, AuthorSchema);
+  const authorIdMap = new Map<string, string>();
+  for (const author of authors) {
+    // Reconstructing normalized author from DB author to then lookup JSON author
+    const normalizedAuthor: NormalizedAuthor = {
+      name: author.author_name ?? null,
+      email: author.email ?? null,
+      orcid: author.orcid ?? null,
+      origin_course: author.origin_course ?? null,
+      id: null,
+    };
+    const jsonAuthor = uniqueAuthors.get(JSON.stringify(normalizedAuthor));
```

Comment:
It's somewhat strange to lookup authors by the stringified object. I suppose this is fine -- is there no other way?